src/omniglot/wrapper.py

```diff
@@ -139,8 +140,182 @@ def run_batches(self, batches, optimizer, train=False, meta_train=False):
+            results.append((trainres, valres))
+        ##
+        results = AggRes(results)
+
+        # Meta gradient step
+        if meta_train:
+            # at the end of collection for K steps N tasks we do the backward
+            # pass.
+            backward(self.meta_loss, self.model.meta_parameters(
+                include_init=False))
+            self._final_meta_update()
+
+        return results
+
+    def run_task(self, task, train, meta_train):
+        """Run model on a given task, first adapting and then evaluating"""
+        self.model.no_collect()
+
+        optimizer = None
+        if train:
+            # TODO: Discuss implementation and correct it.
+            # This line breakes gradient computation for now
+            # meta_layers required_grad properties are set to False if
+            # we call init_adaptation
+            # self.model.init_adaptation()
+            self.model.train()
+
+            optimizer = self.optimizer_cls(
+                self.model.optimizer_parameter_groups(),
+                **self.optimizer_kwargs)
+        else:
+            self.model.eval()
+
+        return self.run_batches(
+            task, optimizer, train=train, meta_train=meta_train)
+
+    def run_batches(self, batches, optimizer, train=False, meta_train=False):
+        """Iterate over task-specific batches.
+
+        Arguments:
+            batches (torch.utils.data.DataLoader): task-specific dataloaders.
+            optimizer (torch.nn.optim): optimizer instance if training is True.
+            train (bool): whether to train on task.
+            meta_train (bool): whether to meta-train on task.
+        """
+        device = next(self.model.parameters()).device
+        self.model.no_collect()
+        res = Res()
+        N = len(batches)
+        for n, (input, target) in enumerate(batches):
+            inner_input = input.to(device, non_blocking=True)
+            inner_target = target.to(device, non_blocking=True)
+
+            # Evaluate model
+            prediction = self.model(inner_input)
+            loss = self.criterion(prediction, inner_target)
+
+            res.log(loss=loss.item(), pred=prediction, target=inner_target)
+
+            # TRAINING #
+            if not train:
+                continue
+
+            final = (n + 1) == N
+            loss.backward()
+
+            if meta_train:
+                opt = SGD(self.model.optimizer_parameter_groups(tensor=True))
+                opt.zero_grad()
+                outer_input, outer_target = next(iter(batches))
+                l_outer, (l_inner, a1, a2) = step(
+                    criterion=self.criterion,
+                    x_inner=inner_input, x_outer=outer_input,
+                    y_inner=inner_target, y_outer=outer_target,
+                    model=self.model,
+                    optimizer=opt, scorer=None)
+                self.meta_loss = self.meta_loss + l_outer
+                del l_inner, a1, a2
```

Comment:
These lines are calculating `outer_loss` at each state of model parameter \theta_{k}^{\tau}. However, I am not sure about how should we handle freezing and unfreezing meta and adaptation layers